src/error.rs

```diff
@@ -23,56 +27,83 @@ use crate::image::ImageFormat;
 ///
 /// This high level enum allows, by variant matching, a rough separation of concerns between
 /// underlying IO, the caller, format specifications, and the `image` implementation.
-#[derive(Debug)]
+#[derive(Snafu, Debug)]
 pub enum ImageError {
     /// An error was encountered while decoding.
     ///
     /// This means that the input data did not conform to the specification of some image format,
     /// or that no format could be determined, or that it did not match format specific
     /// requirements set by the caller.
-    Decoding(DecodingError),
+    ///
+    /// An error was encountered while decoding an image.
+    ///
+    /// This is used as an opaque representation for the [`ImageError::Decoding`] variant. See its
+    /// documentation for more information.
+    ///
+    /// [`ImageError::Decoding`]: enum.ImageError.html#variant.Decoding
+    Decoding { format: ImageFormatHint },
```

Comment:
Is all of this diff necessary for `no_std`? Seems a little weird. In particular the comments are no longer accurate. Enum variants are _not_ opaque. It's definitely preferrable to have opaque types to allow changing the field representation / error messages / internal variants.

Reply:
Yes I think it is...
Previously error types were custom one using `std::error::Error` but that can not work in `no_std`.
The only way to make it work is to use a no_std-compatible lib to handle `Error` and I only found `Snafu`.

But I had a feeling it was messy to bundle the `error.rs` refactor with this PR.
Maybe we can do it in two steps:
- first PR to refactor `error.rs` to use `Snafu`
- "main" PR for `no_std`

It is a big diff but not a complicated one b/c it is mainly code removal:
- no need to `impl Error` when using Snafu
- no need to `impl fmt::Display` when using Snafu

Question: not sure if what I did is a breaking change?

Reply:
These changes to ImageError are a breaking change because the variants of public enums are public. What's not clear to me is whether the changes here are actually necessary. Can Snafu really not derive its trait for enums with tuple variants?

Reply:
What happens if we just gate the `impl std::error::Error` parts behind `std`? Would that cause any problems? Because it would be nice to avoid the code churn if possible

Reply:
Wouldn't that make all the errors not proper errors in `no_std` context?

But we can go with your idea and instead[starting by reversing all changes in `error.rs`]:
- with `std`: `use std::error::Error`, same as before
- `no_std`: `use core::error::Error`
   That would make `no_std` builds require nightly for now; but that is fine for my use case.

Would that be ok for you?

For completeness: there is also the "duplicated code" alternative:
- move original `error.rs` -> `error/error_compat.rs`
- rename this forks `error.rs` -> `error/error_new.rs`
- in `error/mod.rs` switch impl based on std vs no_std